File: Assign_3/classification.py
import math
from collections import Counter
from typing import List, Dict, Tuple, Optional

class Matrix:
    def __init__(self, data_file: str = None, data: List[List[str]] = None):
        self.matrix = []
        if data_file:
            with open(data_file, 'r') as f:
                for line in f:
                    row = line.strip().split()
                    if row:
                        self.matrix.append(row)
        elif data:
            self.matrix = data

    def element(self, i: int, j: int) -> str:
        return self.matrix[i][j]

    def size_x(self) -> int:
        return len(self.matrix[0]) if self.matrix else 0

    def size_y(self) -> int:
        return len(self.matrix)

    def get_attributes(self) -> List[str]:
        return self.matrix[0][:-1] if self.matrix else []

    def get_attributes_values(self) -> Dict[str, List[str]]:
        attributes_values = {}
        for j in range(self.size_x()):
            values = [self.matrix[i][j] for i in range(1, self.size_y())]
            attributes_values[self.matrix[0][j]] = sorted(list(set(values)))
        return attributes_values

    def get_attribute_values(self, attribute: str) -> List[str]:
        return self.get_attributes_values().get(attribute, [])

    def get_score_range(self) -> List[str]:
        return self.get_attribute_values(self.matrix[0][-1])

    def attribute_index(self, attribute: str) -> int:
        return self.matrix[0].index(attribute)

    def get_attribute_values_scores(self, attribute: str) -> Dict[str, List[str]]:
        idx = self.attribute_index(attribute)
        values_scores = {}
        for value in self.get_attribute_values(attribute):
            scores = []
            for i in range(1, self.size_y()):
                if self.matrix[i][idx] == value:
                    scores.append(self.matrix[i][-1])
            values_scores[value] = scores
        return values_scores

    def get_scores(self) -> List[str]:
        return [row[-1] for row in self.matrix[1:]]

    def filter_matrix(self, attribute: str, value: str) -> 'Matrix':
        idx = self.attribute_index(attribute)
        # Keep the header row and filter the data rows
        new_matrix = [self.matrix[0][:idx] + self.matrix[0][idx+1:]]  # Keep header
        for row in self.matrix[1:]:  # Skip header
            if row[idx] == value:
                new_matrix.append(row[:idx] + row[idx+1:])
        return Matrix(data=new_matrix)

    def display(self):
        for row in self.matrix:
            print(' '.join(row))

def get_unique_scores(scores: List[str]) -> List[str]:
    return sorted(list(set(scores)))

def get_frequent_score(scores: List[str]) -> str:
    return Counter(scores).most_common(1)[0][0]

def compute_entropy(scores: List[str]) -> float:
    if not scores:
        return 0.0
    
    score_counts = Counter(scores)
    total = len(scores)
    entropy = 0.0
    
    for count in score_counts.values():
        p = count / total
        entropy -= p * math.log2(p)
    
    return entropy

def compute_attribute_entropy_gain(matrix: Matrix, attribute: str) -> float:
    scores = matrix.get_scores()
    original_entropy = compute_entropy(scores)
    
    values_scores = matrix.get_attribute_values_scores(attribute)
    after_entropy = 0.0
    
    for value_scores in values_scores.values():
        p = len(value_scores) / len(scores)
        after_entropy += p * compute_entropy(value_scores)
    
    return original_entropy - after_entropy

class Tree:
    def __init__(self):
        self.node = ""
        self.branch = ""
        self.children = []

    def build_tree(self, matrix: Matrix) -> 'Tree':
        scores = matrix.get_scores()
        unique_scores = get_unique_scores(scores)
        
        if len(unique_scores) == 1:
            self.node = unique_scores[0]
            return self
        
        if matrix.size_x() == 1:
            self.node = get_frequent_score(scores)
            return self
        
        attributes = matrix.get_attributes()
        max_gain = -1.0  # Initialize to negative value
        max_attribute = ""
        
        for attribute in attributes:
            gain = compute_attribute_entropy_gain(matrix, attribute)
            if gain > max_gain:
                max_gain = gain
                max_attribute = attribute
        
        if max_attribute == "":  # No attribute with positive gain
            self.node = get_frequent_score(scores)
            return self
            
        self.node = max_attribute
        values = matrix.get_attribute_values(max_attribute)
        
        for value in values:
            new_matrix = matrix.filter_matrix(max_attribute, value)
            child = Tree()
            child.branch = value
            
            if new_matrix.size_x() == 1 or len(new_matrix.get_scores()) == 0:
                child.node = get_frequent_score(scores)  # Use parent's scores for empty matrix
            else:
                child.build_tree(new_matrix)
            
            self.children.append(child)
        
        return self

    def print_tree(self, depth: int = -1):
        indent = "\t" * max(0, depth)
        if self.branch:
            print(f"{indent}{self.branch}")
            indent += "\t"
        if depth == -1 and self.branch:
            print("\t", end="")
        print(f"{indent}{self.node}")
        for child in self.children:
            child.print_tree(depth + 1)

    def test_tree(self, matrix: Matrix) -> List[str]:
        attributes = matrix.get_attributes()
        score_range = matrix.get_score_range()
        test_scores = []
        
        for i in range(1, matrix.size_y()):
            values = [matrix.element(i, j) for j in range(len(attributes))]
            prediction = self.temp_test_tree(attributes, values, score_range)
            test_scores.append(prediction)
        
        return test_scores

    def temp_test_tree(self, attributes: List[str], values: List[str], score_range: List[str]) -> str:
        # If current node is a leaf node (contains a score)
        if self.node in score_range:
            return self.node
        
        # Find the matching attribute
        for i, attribute in enumerate(attributes):
            if self.node == attribute:
                # Find the matching branch for this attribute value
                for child in self.children:
                    if child.branch == values[i]:
                        # If child node is a leaf node
                        if child.node in score_range:
                            return child.node
                        
                        # If not a leaf node, recursively traverse the tree
                        new_attributes = [a for j, a in enumerate(attributes) if j != i]
                        new_values = [v for j, v in enumerate(values) if j != i]
                        return child.temp_test_tree(new_attributes, new_values, score_range)
                
                # If no matching branch is found, return the most common score
                return get_frequent_score(score_range)
        
        # Default case: return the first score in the range
        return score_range[0] if score_range else "unknown"

def main():
    # Load training data
    train_matrix = Matrix("Train.dat")
    root = Tree().build_tree(train_matrix)
    
    print("Decision Tree Structure:")
    root.print_tree()

    # Load test data
    test_matrix = Matrix("Test.dat")
    test_scores = root.test_tree(test_matrix)
    original_scores = test_matrix.get_scores()

    print("\nOriginal_Scores (from Test.dat):")
    print("  ".join(original_scores))
    
    print("\nPredicted_Scores (from Test.dat):")
    print("  ".join(test_scores))
    
    # Calculate accuracy
    correct = sum(1 for t, o in zip(test_scores, original_scores) if t == o)
    accuracy = 100.0 * correct / len(test_scores)
    print(f"\nAccuracy on test data: {accuracy}%")

if __name__ == "__main__":
    main() 

File: Assign_3/classification.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <sstream>
#include <vector>
#include <map>
#include <algorithm>
#include <cmath>
#include <numeric>  // Add this for inner_product
using namespace std;

class MatrixCls {
private:
    vector<vector<string>> Matrix;
public:
    MatrixCls(string Data_File) {
        ifstream Data(Data_File);
        string line, item;
        vector<string> Row;
        while(getline(Data, line)) {
            istringstream iss(line);
            while(iss >> item && item.length()) Row.push_back(item);
            if(line.length()) {
                Matrix.push_back(Row);
                Row.clear();
            }
        }
        Data.close();
    }

    MatrixCls(vector<vector<string>> A_Matrix) : Matrix(A_Matrix) {}
    MatrixCls() {}
    ~MatrixCls() {}

    string Element(int i, int j) const { return Matrix[i][j]; }
    int SizeX() const { return Matrix[0].size(); }
    int SizeY() const { return Matrix.size(); }

    vector<string> GetAttributes() const {
        return vector<string>(Matrix[0].begin(), Matrix[0].end() - 1);
    }

    map<string, vector<string>> GetAttributesValues() const {
        map<string, vector<string>> Attributes_Values;
        for(int j = 0; j < SizeX(); j++) {
            vector<string> values;
            for(int i = 1; i < SizeY(); i++) values.push_back(Matrix[i][j]);
            sort(values.begin(), values.end());
            values.erase(unique(values.begin(), values.end()), values.end());
            Attributes_Values[Matrix[0][j]] = values;
        }
        return Attributes_Values;
    }

    vector<string> GetAttributeValues(string The_Attribute) const {
        return GetAttributesValues()[The_Attribute];
    }

    vector<string> GetScoreRange() const {
        return GetAttributesValues()[Matrix[0].back()];
    }

    int AttributeIndex(string The_Attribute) const {
        auto it = find(Matrix[0].begin(), Matrix[0].end(), The_Attribute);
        return it - Matrix[0].begin();
    }

    map<string, vector<string>> GetAttributeValuesScores(string The_Attribute) const {
        int idx = AttributeIndex(The_Attribute);
        map<string, vector<string>> result;
        auto values = GetAttributeValues(The_Attribute);
        for(const auto& val : values) {
            vector<string> scores;
            for(int i = 1; i < SizeY(); i++)
                if(Matrix[i][idx] == val)
                    scores.push_back(Matrix[i].back());
            result[val] = scores;
        }
        return result;
    }

    vector<string> GetScores() const {
        vector<string> scores;
        for(int i = 1; i < SizeY(); i++)
            scores.push_back(Matrix[i].back());
        return scores;
    }

    MatrixCls operator()(const MatrixCls& A_Matrix, string The_Attribute, string The_Value) {
        int idx = A_Matrix.AttributeIndex(The_Attribute);
        vector<vector<string>> newMatrix;
        vector<string> header;
        
        for(int j = 0; j < A_Matrix.SizeX(); j++)
            if(j != idx) header.push_back(A_Matrix.Element(0,j));
        if(!header.empty()) newMatrix.push_back(header);

        for(int i = 1; i < A_Matrix.SizeY(); i++) {
            if(A_Matrix.Element(i,idx) == The_Value) {
                vector<string> row;
                for(int j = 0; j < A_Matrix.SizeX(); j++)
                    if(j != idx) row.push_back(A_Matrix.Element(i,j));
                if(!row.empty()) newMatrix.push_back(row);
            }
        }
        Matrix = newMatrix;
        return *this;
    }

    void Display() const {
        for(const auto& row : Matrix) {
            for(const auto& elem : row) cout << " " << elem;
            cout << endl;
        }
    }
};

vector<string> GetUniqueScores(vector<string> Scores) {
    sort(Scores.begin(), Scores.end());
    Scores.erase(unique(Scores.begin(), Scores.end()), Scores.end());
    return Scores;
}

string GetFrequentScore(const vector<string>& Scores) {
    map<string,int> freq;
    for(const auto& score : Scores) freq[score]++;
    return max_element(freq.begin(), freq.end(),
        [](const pair<string,int>& a, const pair<string,int>& b) {
            return a.second < b.second;
        })->first;
}

double ComputeEntropy(const vector<string>& Scores) {
    if(Scores.empty()) return 0.0;
    
    map<string,int> freq;
    for(const auto& score : Scores) freq[score]++;
    
    double entropy = 0.0;
    double size = Scores.size();
    for(const auto& pair : freq) {
        double p = pair.second / size;
        entropy -= p * log2(p);
    }
    return entropy;
}

double ComputeAttributeEntropyGain(const MatrixCls& Remain_Matrix, string The_Attribute) {
    auto scores = Remain_Matrix.GetScores();
    double originalEntropy = ComputeEntropy(scores);
    
    auto valuesScores = Remain_Matrix.GetAttributeValuesScores(The_Attribute);
    double afterEntropy = 0.0;
    for(const auto& pair : valuesScores) {
        afterEntropy += ComputeEntropy(pair.second) * pair.second.size() / scores.size();
    }
    return originalEntropy - afterEntropy;
}

class Tree {
public:
    string Node, Branch;
    vector<Tree*> Child;
    Tree() : Node(""), Branch("") {}
    
    string Temp_TestTree(Tree* tree, const vector<string>& Attributes, 
                        const vector<string>& Values, const vector<string>& Score_Range) {
        auto it = find(Score_Range.begin(), Score_Range.end(), tree->Node);
        if(it != Score_Range.end()) return tree->Node;

        auto attrIt = find(Attributes.begin(), Attributes.end(), tree->Node);
        if(attrIt != Attributes.end()) {
            int idx = attrIt - Attributes.begin();
            for(auto child : tree->Child) {
                if(child->Branch == Values[idx]) {
                    if(find(Score_Range.begin(), Score_Range.end(), child->Node) != Score_Range.end())
                        return child->Node;
                        
                    vector<string> newAttrs, newVals;
                    for(size_t i = 0; i < Attributes.size(); i++)
                        if(i != idx) {
                            newAttrs.push_back(Attributes[i]);
                            newVals.push_back(Values[i]);
                        }
                    return Temp_TestTree(child, newAttrs, newVals, Score_Range);
                }
            }
            return GetFrequentScore(Score_Range);
        }
        return Score_Range.empty() ? "unknown" : Score_Range[0];
    }

    Tree* BuildTree(Tree* tree, const MatrixCls& Remain_Matrix) {
        if(!tree) tree = new Tree();
        
        auto scores = Remain_Matrix.GetScores();
        auto uniqueScores = GetUniqueScores(scores);
        
        if(uniqueScores.size() == 1) {
            tree->Node = uniqueScores[0];
            return tree;
        }
        
        if(Remain_Matrix.SizeX() == 1) {
            tree->Node = GetFrequentScore(scores);
            return tree;
        }
        
        auto attributes = Remain_Matrix.GetAttributes();
        string maxAttr;
        double maxGain = 0;
        
        for(const auto& attr : attributes) {
            double gain = ComputeAttributeEntropyGain(Remain_Matrix, attr);
            if(gain > maxGain) {
                maxGain = gain;
                maxAttr = attr;
            }
        }
        
        tree->Node = maxAttr;
        auto values = Remain_Matrix.GetAttributeValues(maxAttr);
        
        for(const auto& val : values) {
            MatrixCls newMatrix;
            newMatrix = newMatrix(Remain_Matrix, maxAttr, val);
            Tree* child = new Tree();
            child->Branch = val;
            
            if(newMatrix.SizeX() == 1)
                child->Node = GetFrequentScore(newMatrix.GetScores());
            else
                BuildTree(child, newMatrix);
                
            tree->Child.push_back(child);
        }
        return tree;
    }

    void PrintTree(Tree* tree, int depth = -1) {
        string indent(max(0, depth), '\t');
        if(!tree->Branch.empty()) {
            cout << indent << tree->Branch << endl;
            indent += "\t";
        }
        if(depth == -1 && !tree->Branch.empty()) cout << "\t";
        cout << indent << tree->Node << endl;
        for(auto child : tree->Child) PrintTree(child, depth + 1);
    }

    vector<string> TestTree(Tree* tree, const MatrixCls& The_Matrix) {
        vector<string> results;
        auto attrs = The_Matrix.GetAttributes();
        auto range = The_Matrix.GetScoreRange();
        
        for(int i = 1; i < The_Matrix.SizeY(); i++) {
            vector<string> values;
            for(int j = 0; j < attrs.size(); j++)
                values.push_back(The_Matrix.Element(i,j));
            results.push_back(Temp_TestTree(tree, attrs, values, range));
        }
        return results;
    }
};

int main() {
    MatrixCls TrainMatrix("Train.dat");
    Tree* root = new Tree();
    root = root->BuildTree(root, TrainMatrix);
    
    cout << "Decision Tree Structure:" << endl;
    root->PrintTree(root);

    MatrixCls TestMatrix("Test.dat");
    auto testScores = root->TestTree(root, TestMatrix);
    auto origScores = TestMatrix.GetScores();

    cout << "\nOriginal_Scores (from Test.dat):\n";
    for(const auto& score : origScores) cout << score << "  ";
    
    cout << "\n\nPredicted_Scores (from Test.dat):\n";
    for(const auto& score : testScores) cout << score << "  ";
    
    // Calculate accuracy using a simple loop
    int correct = 0;
    for(size_t i = 0; i < testScores.size(); i++) {
        if(testScores[i] == origScores[i]) {
            correct++;
        }
    }
    
    double accuracy = 100.0 * correct / testScores.size();
    cout << "\n\nAccuracy on test data: " << accuracy << "%" << endl;
    
    delete root;
}